### ISSUE TITLE
htseq-count python tutorial attribute counts error

### ISSUE THREAD

**Issue by Jacobvazquez18** [htseq/htseq#23]
Hello,

I'm following the htseq-count tutorial for RNA-seq (counting the overlapping genes and exons) here https://htseq.readthedocs.io/en/master/tour.html. However, when I get to the point where I need to find the overlaps in the .sam file and .gtf file, I get an error. This is the code I ran originally that gave the error:

`## find overlaps of genes and exons using sam file and .gtf file
for alnmt in sam_file:
    if alnmt.aligned:
        iset = None
        for iv2, step_set in exons[alnmt.iv].steps():
            if iset is None:
                iset = step_set.copy()
            else:
                iset.intersection_update(step_set)
        

And this is the error message after running:

---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-22-1071efdd4856> in <module>
      3     if alnmt.aligned:
      4         iset = None
----> 5         for iv2, step_set in exons[alnmt.iv].steps():
      6             if iset is None:
      7                 iset = step_set.copy()

~/anaconda3/lib/python3.8/site-packages/HTSeq/_HTSeq_internal.py in ChromVector_steps(cv)
     31             )
     32     elif isinstance(cv.array, HTSeq.StepVector.StepVector):
---> 33         for start, stop, value in cv.array[cv.iv.start:cv.iv.end].get_steps():
     34             yield (HTSeq.GenomicInterval(
     35                 cv.iv.chrom, start, stop, cv.iv.strand), value,

~/anaconda3/lib/python3.8/site-packages/HTSeq/StepVector.py in get_steps(self, values_only, merge_steps)
    407       startvals = self._swigobj.get_values_pystyle( self.start )
    408       prevstart = self.start
--> 409       prevval = startvals.next().second
    410       for pair in startvals:
    411          stepstart, value = pair.first, pair.second

AttributeError: '_StepVector_Iterator_obj' object has no attribute 'next'

It said "object has no attribute 'next'". Therefore, I looked back at the StepVector.py script and found that the method is called "get_steps()" and not "steps() as it was written in the tutorial code I used above. Therefore, I changed it to "get_steps()" and ran it but got the following error:

---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-23-4a780ddcd656> in <module>
      3     if alnmt.aligned:
      4         iset = None
----> 5         for iv2, step_set in exons[alnmt.iv].get_steps():
      6             if iset is None:
      7                 iset = step_set.copy()

AttributeError: 'HTSeq._HTSeq.ChromVector' object has no attribute 'get_steps'

Any idea as to what's going on and how I can work around it?

**Comment by Irnov1** [comment]
I am having the same issue running HTSeq. My code used to work, but I recently upgraded my python, conda, etc. I am running this on Mac OS Big Sur with python 3.8.8, htseq 0.13.5. Any thoughts on how to get around this issue?

**Comment by iosonofabio** [comment]
I am a little busy as of lately, but the error message seems clear enough to start a little debugging session... You tried `next(x)` like in python3 lingo instead of `x.next()`?

**Comment by Irnov1** [comment]
Yeah, I tried modifying line 409 in 'StepVector.py' from: `prevval = startvals.next().second ` to: `prevval = next(startvals).second`, but still getting the same error (see below)

------------------------------
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-120-d11ed4e052e9> in <module>
     23 cv = cvg[HTSeq.GenomicInterval('NC_001318.1_chr',0,size,'.' )]
     24 
---> 25 for iv, value in cv.steps(): print(value)

~/opt/anaconda3/lib/python3.8/site-packages/HTSeq/_HTSeq_internal.py in ChromVector_steps(cv)
     31             )
     32     elif isinstance(cv.array, HTSeq.StepVector.StepVector):
---> 33         for start, stop, value in cv.array[cv.iv.start:cv.iv.end].get_steps():
     34             yield (HTSeq.GenomicInterval(
     35                 cv.iv.chrom, start, stop, cv.iv.strand), value,

~/opt/anaconda3/lib/python3.8/site-packages/HTSeq/StepVector.py in get_steps(self, values_only, merge_steps)
    407       startvals = self._swigobj.get_values_pystyle( self.start )
    408       prevstart = self.start
--> 409       prevval = next(startvals).second
    410       for pair in startvals:
    411          stepstart, value = pair.first, pair.second

AttributeError: '_StepVector_Iterator_int' object has no attribute 'next'

**Comment by iosonofabio** [comment]
Alright guys, thre's a few things:

1. The tutorial code is actually run on CI so it *must* work when compiling from source correctly.
2. Indeed, I just cloned the git repo, ran `python setup.py build`, added the right folder to my `PYTHONPATH`, and I can run the `tour.rst` tutorial code just fine
3. `exons[iv]` is obviously **not** a `StepVector. How do I know? Well...

```python
 print(exons[iv])
<ChromVector object, III:[23850,23950)/., step>
```

It is a `ChromVector` instead. So looking for methods in `StepVector.py` wouldn't be so useful. `ChromeVector`, on the other hand, is a Cython object sitting in `src/HTSeq/_HTSeq.pyx` and has the `steps` method (of course, otherwise how could it work on the CI computer?). If you look at it, it says:

```python
def steps(self):
    return _HTSeq_internal.ChromVector_steps(self)
```

so even though it's Cython, this particular method is implemented in pure Python for convenience. So you can open `HTSeq/_HTSeq_internal.py` and you'lll find that function happily sitting there:

```python
def ChromVector_steps(cv):
...

and it's a pure Python generator that uses the `StepVector.get_steps` function under the hood in some circumstances. So back to the Cython file we are brought.

So in conclusion:
- `StepVector` does not have `steps` and was never designed to, so don't look for it ;-)
- Your Cython compilation seems to be a little odd, try to install from the master branch and see if it works

**Comment by Irnov1** [comment]
Hi Fabio, thank you so much for the pointers! I decided to just re-install HTSeq using pip this time (last time was through conda) and it's working well now! The error was likely due to the version in conda repository. Thanks again!

**Comment by iosonofabio** [comment]
Great to hear. Yes following the actual install instructions helps ;-)
